### PR TITLE
simple, single retry for panos' commit

### PR DESCRIPTION
**Pull request by eikenb**
Recommendation at this point was to retry only once after a decent delay
based on SE feedback.

Retry is very simple and groups together the commit and wait calls as
either could fail (both make http calls to the panos API with failure
conditions).

Added checks to existing tests as they already provided decent coverage.

--

PR notes:
1) what do you think of the the retry tests using expectedRetries() vs. including the expected retries in the table data?

### CODE REVIEW

File: handler/panos_test.go [hunk 1]
```diff
@@ -206,18 +206,39 @@ func TestPanosCommit(t *testing.T) {
...
+				case c && w: // everyone happy
+					return 1, 1
+				case c && !w: // wait fails
+					return 2, 2
+				}
+				return -1, -1 // never reached
+			}
+			commitTries, waitTries := expectedRetries()
+			m.AssertNumberOfCalls(t, "Commit", commitTries)
+			m.AssertNumberOfCalls(t, "WaitForJob", waitTries)
```
Comment (lornasong) [inline handler/panos_test.go:241]: I've never used this! Good to know about!

**Review by lornasong — APPROVED** [review]
looks good to me! the changes are very readable, particularly the switch-statement and comments in the tests

File: handler/panos.go [hunk 2]
```diff
@@ -138,18 +144,29 @@ func (h *Panos) commit() error {
...
 	}
-	if err != nil {
-		log.Printf("[ERR] (handler.panos) error committing: %s. Server response: '%s'", err, resp)
-		return err
-	}
 
-	if err := h.client.WaitForJob(job, nil); err != nil {
-		log.Printf("[ERR] (handler.panos) error waiting for panos commit to finish: %s", err)
-		return err
+	if err := tryCommit(); err != nil {
```
Comment (findkim) [inline handler/panos.go:164]: Not sure if you're already aware, but wanted mention that there's an existing [`retry` package](https://github.com/hashicorp/consul-terraform-sync/blob/master/retry/retry.go) that is pretty small and is currently used for [retrying task execution in daemon mode](https://github.com/hashicorp/consul-terraform-sync/blob/master/controller/readwrite.go#L203). It has slightly different logic in that it uses a random delay based on seed. Would you want to repurpose/modifie for this single retry too?

Comment (eikenb) [reply handler/panos.go:164]: I missed retry getting added. I had thought about using a more standard retry algorithm but was waiting to talk to Tristan about it. Now (post meeting) that I have and determined exponential backoff is appropriate I can just use it as is... Thanks!

Also, more importantly IMO, that reminded me we really should be passing the Context through the handler Do() calls to allow for interruptions.

I'm going to rework handlers to pass Context and this case to use that retry package.

Comment (eikenb) [reply handler/panos.go]: All done with these changes. Thanks again.

File: handler/panos.go [hunk 3]
```diff
@@ -19,6 +21,9 @@ const (
 	// Users with custom roles currently return an error for an empty commit with
 	// this server response prefix. See GH-73 for more details.
 	emptyCommitServerRespPrefix = `<response status="success" code="13">`
+
+	// max number of retries
+	maxRetris = 4
```
Comment (lornasong) [inline handler/panos.go:26]: small typo: "maxRetris" => "maxRetries"

File: retry/retry.go [hunk 4]
```diff
@@ -25,7 +27,7 @@ func NewRetry(maxRetry uint, seed int64) Retry {
 }
 
 // Do calls a function with exponential retry with a random delay.
-func (r *Retry) Do(ctx context.Context, f func(context.Context) error, desc string) error {
+func (r Retry) Do(ctx context.Context, f func(context.Context) error, desc string) error {
```
Comment (lornasong) [inline retry/retry.go:30]: ahhh👍

File: retry/retry.go [hunk 5]
```diff
@@ -81,3 +86,11 @@ func (r *Retry) waitTime(attempt uint) int {
 	total := (baseTimeSeconds + delay) * float64(time.Second)
 	return int(total)
 }
+
+// Test version, returns retry in test mode (nanosecond retry delay).
+func NewTestRetry(maxRetry uint) Retry {
```
Comment (lornasong) [inline retry/retry.go:91]: good idea! (note to self to use this in the future)

**Review by lornasong — COMMENTED** [review]
this looks good to me! 🎉 reusing the retry library as @findkim mentioned. i only noticed a very minor typo that we can include as part of some other pr for the next time we work in the panos code area

File: handler/panos.go (hunk 3 above)
Comment (eikenb) [reply handler/panos.go:26]: Saw the approved and missed this. Will fix in a small follow up commit.

File: retry/retry.go (hunk 5 above)
Comment (eikenb) [reply retry/retry.go:91]: I considered a 'Retrier' interface that I could swap out with a test version but the semantics of the interface meant the nil value resulted in a panic and the changes to fix that were ugly. So I preserved using the struct directly and added a test mode to it.

**Pull request merged**